lambda/matching.ml

```diff
@@ -1348,68 +1444,119 @@ and precompile_var args cls def k =
+              extract_equiv_head Half_simple.to_pattern orp rem
+            in
+            let orpm =
+              { cases =
+                  (patl, action)
+                  :: List.map (fun ((_, ps), action) -> (ps, action)) others;
+                args =
+                  ( match args with
+                  | _ :: r -> r
+                  | _ -> assert false
+                  );
+                default = Default_environment.pop_compat orp def
+              }
+            in
+            let pm_fv = pm_free_variables orpm in
+            let vars =
+              (* bound variables of the or-pattern and used in the orpm actions *)
+              Typedtree.pat_bound_idents_full orp
+              |> List.filter (fun (id, _, _) -> Ident.Set.mem id pm_fv)
+              |> List.map (fun (id, _, ty) ->
+                     (id, Typeopt.value_kind orp.pat_env ty))
+            in
+            let or_num = next_raise_count () in
+            let new_patl = Parmatch.omega_list patl in
+            let mk_new_action ~vars =
+              Lstaticraise (or_num, List.map (fun v -> Lvar v) vars)
+            in
+            let rem_cases, rem_handlers = do_cases rem in
+            let cases =
+              Simple.explode_or_pat (p, new_patl) ~arg:argo
+                ~mk_action:mk_new_action ~vars:(List.map fst vars) rem_cases
+            in
+            let handler =
+              { provenance = [ [ orp (* FIXME? *) ] ];
+                exit = or_num;
+                vars;
+                pm = orpm
+              }
+            in
+            (cases, handler :: rem_handlers)
+      )
   in
   let cases, handlers = do_cases ors in
-  let matrix = as_matrix (cls @ ors)
+  let matrix =
+    as_matrix
+      (fun x -> x)
+      (List.map (fun ((p, ps), act) -> ((Simple.to_pattern p, ps), act)) cls
+      @ List.map
+          (fun ((p, ps), act) -> ((Half_simple.to_pattern p, ps), act))
+          ors
+      )
   and body = { cases = cls @ cases; args; default = def } in
   ( { me = PmOr { body; handlers; or_matrix = matrix };
       matrix;
       top_default = def
     },
     k )
 
+let split_and_precompile_nonempty argo pm =
+  let pm =
+    { pm with cases = List.map (Half_simple.of_clause ~args:pm.args) pm.cases }
+  in
+  let { me = next }, nexts = split_or argo pm.cases pm.args pm.default in
+  if
+    dbg
+    && (nexts <> []
+       ||
+       match next with
+       | PmOr _ -> true
+       | _ -> false
+       )
+  then (
+    Format.eprintf "** SPLIT **\n";
+    pretty_hc_pm pm;
+    pretty_precompiled_res next nexts
+  );
+  (next, nexts)
+
+let split_and_precompile_simplified pm =
+  let { me = next }, nexts = split_no_or pm.cases pm.args pm.default [] in
+  if
+    dbg
+    && (nexts <> []
+       ||
+       match next with
+       | PmOr _ -> true
+       | _ -> false
+       )
+  then (
+    Format.eprintf "** SPLIT **\n";
+    pretty_sc_pm pm;
+    pretty_precompiled_res next nexts
+  );
+  (next, nexts)
+
```

Comment:
The debug part in `split_and_precompile_` is really noisy. I think it should be factorized, otherwise the small difference between the three variants are lost in the noise.

Reply:
https://github.com/ocaml/ocaml/pull/9321/commits/36961ba085d8ccf9832f24c597257471056fb01d does this I believe.

Reply:
@Octachron remarked that this is duplication we introduced in this PR, so it would be better fixed now. I added two more commits to do it.